File: neetcode150/sliding_window/leetcode239.py
from collections import deque
from typing import List


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        indices = deque()
        result = []

        for end, num in enumerate(nums):
            # keep track of current window
            # if front index is not within the window, we can get rid of it
            while indices and indices[0] <= end - k:
                indices.popleft()

            # if the last item in the index is less than the current num
            # then we can get rid of it
            while indices and nums[indices[-1]] <= num:
                indices.pop()

            indices.append(end)

            # once we have a window of size k, we can start adding values
            # to our result list
            if end - k + 1 >= 0:
                result.append(nums[indices[0]])

        return result

File: js_algorithms/neetcode150/sliding_window/leetcode239.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
const maxSlidingWindow = function (nums, k) {
  // keep track of indices of certain numbers
  const deque = [];
  const result = [];

  for (let i = 0; i < nums.length; i++) {
    // if i - last > k, then its out of the window
    while (deque.length && i - deque[0] >= k) {
      deque.shift();
    }

    while (deque.length && nums[deque[deque.length - 1]] < nums[i]) {
      deque.pop();
    }

    deque.push(i);

    if (i >= k - 1) {
      result.push(nums[deque[0]]);
    }
  }

  return result;
};

module.exports = maxSlidingWindow;
